flow/project.py

```diff
@@ -1490,7 +1491,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
             Show a progress bar during execution.
         :type progess:
             bool
+        :param order:
+            Specify the order of operations, possible values are:
+                * None or 'none' (no reordering, do not assume any order)
```

Comment:
I'd prefer to specify what this does instead of what it does not do. Something like "uses order as returned by (some function)" or some definition from the FlowProject. I'm not sure how this works internally, but I find the current docstring unclear and would want to search the source code for clarification of what "no assumed order" actually means.

Reply:
Like I explained in my response to @b-butler , I'd prefer to not explain what the code does "implicitly" in order to avoid users starting to rely on that. I think it would be highly preferable for us to have the option to change the default behavior at any point in time without needing to go through a deprecation cycle.

This is similar to the debate about Python's built-in `dict` class, where the insertion order was considered implementation detail up until Python 3.7, although it has been identical to `OrderedDict` for quite a while now. See this [StackOverflow reply](https://stackoverflow.com/a/39980744) for some details. I don't think it is a good idea for us to promise any kind of order while the implementation is still highly in flux.

Reply:
I guess I didn’t realize this was the same as @b-butler’s suggestion. As a new user I would have difficulty understanding the default behavior as it is. I think it would be fine to make by-job the default and even change the execution order without a full deprecation, if you view the API guarantee as the end result of the state machine instead of the exact order of execution. However, I understand your reasoning and am fine with your decisions here.